crypto/src/key/hdkd/mod.rs

```diff
@@ -0,0 +1,281 @@
+// Copyright (c) 2022 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::key::hdkd::DerivationError::*;
+use crate::key::rschnorr::MLRistrettoPrivateKey;
+use crate::key::{PrivateKey, PrivateKeyHolder};
+use core::fmt;
+use schnorrkel::derive::{ChainCode, CHAIN_CODE_LENGTH};
+use schnorrkel::ExpansionMode::Ed25519;
+use std::fmt::Formatter;
+use std::str::FromStr;
+
+const PREFIX: &str = "m";
+const SEPARATOR: &str = "/";
+const HARDENED_BIT: u32 = 0x80000000;
+
+#[derive(thiserror::Error, Debug, PartialEq, Eq)]
+#[non_exhaustive]
+pub enum DerivationError {
+    #[error("Initialization error")]
+    FailedToInitializeChainstate(String),
+    #[error("Invalid child number {0}")]
+    InvalidChildNumber(u32),
+    #[error("Malformed child number format")]
+    InvalidChildNumberFormat,
+    #[error("Malformed derivation path format")]
+    InvalidDerivationPathFormat,
+    #[error("Unsupported key type")]
+    UnsupportedKeyType,
+    #[error("Unsupported child number type")]
+    UnsupportedChildNumberType,
+}
+
+/// BIP32-like child numbers. Currently we only support hardened derivations
+#[derive(Debug, Clone, PartialEq, Eq, Ord, PartialOrd, Hash)]
+pub enum ChildNumber {
+    Hardened(u32),
+}
+
+impl ChildNumber {
+    /// Return a hardened child number
+    pub fn hardened(index: u32) -> Result<Self, DerivationError> {
```

Comment:
Like discussion in the call, index can mean index and can mean index | HARDENED_BIT, please disambiguate them.